sagemathinc/cocalc
# webapp/kiosk: remotely control an embedded cocalc instance

# Description

just a POC, we'll discuss this later. auxiliary files are here: https://github.com/haraldschilly/cocalc-embedding-test/ (hosted on my personal domain)

* extra work:
   - [x] merge conflict
   - [x] opening a file with a direct path no longer works

# Testing Steps
* probably the most important aspect is that this doesn't impact normal usage
* it typescripts these server settings, hence they should still work
* opening a link for fullscreen=kiosk mode should open that file and nothing else (can be tested by shift clicking on a file tab)
* fullscreen kiosk mode has a couple of modifications.
  *  compared to what's in master, this adds a bit of logic to deal with switching projects by merging what's in the store, etc. I think I got it, and from my tests it all works, but this definitely needs a second pair of eyes. i.e. projects.cjsx → projects table → _change.
  * there is also a `load_single_project` function, which goes a step further from what's there before. in turn, `load_recent_projects()` is no longer called at all.
  * no session management, no point in that
  * the app_shared.cjsx logic for the active app content is also more restrictive. i.e. you can only see a file or a cocalc banner (no file listing, etc.)
* the actual meat is this iframe-communication file. it basically receives and checks json objects from a parent page. then it dispatches commands. the security aspect is that it only listens to https and a configured list of domains. that's explained in that new server setting config. there is also a test file for that. so, for now I'm for being restrictive. once we know this is really safe, we can open this up to anyone, I guess.
* I've added you to two of my test projects in "test". so, as long as you're signed in, you can open [this page](https://dev.harald.schil.ly/cocalc-embedding-test/embedding-test.html) and click the buttons. e.g.
  *  you can try right after refreshing the page to quickly click on "status". first, no response, then it tells you it is "connecting" and then "connected".
  * another neat thing to try is clicking "halt" in the jupyter notebook

… and I noticed there is a new test failure after merging with master. So, that's something i'll take care of tomorrow.



![2019-08-01-cross-domain-comm](https://user-images.githubusercontent.com/207405/62298021-7abd2b80-b472-11e9-9bad-7202f60ce0ca.png)


### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):

Front end:
